### PR TITLE
Nft benchmarks

### PR DESCRIPTION
**Pull request by jordy25519**
Add benchmarking
Some weights could be modified to scale with the input to avoid over-charging

### CODE REVIEW

File: crml/nft/src/types.rs [hunk 1]
```diff
@@ -101,18 +101,9 @@ impl NFTAttributeValue {
 	}
 }
 
-/// A single data field of an NFT
```
Comment (jordy25519) [inline crml/nft/src/types.rs:104]: unused

File: crml/nft/src/benchmarking.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+fn setup_collection<T: Trait>(creator: T::AccountId) -> (CollectionId, NFTSchema, RoyaltiesSchedule<T::AccountId>) {
+	// Collection id with max length
+	let collection_id = [1_u8; MAX_COLLECTION_ID_LENGTH as usize].to_vec();
+
+	// Schema with max. attributes
+	let schema = (0..MAX_SCHEMA_FIELDS)
+		.map(|i| {
+			let mut v = b"probably-a-decent-upper-length-for-an-attribte-name".to_vec();
+			v.push(i as u8);
+			(v, NFTAttributeValue::String(Default::default()).type_id())
```
Comment (aliXsed) [inline crml/nft/src/benchmarking.rs:40]: I would suggest a String longer than Default::default() for benchmarking here. Something that would satisfy T::MaxAttributeLength::get()

File: crml/nft/src/benchmarking.rs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+	}
+
+	create_token {
+		let creator: T::AccountId = whitelisted_caller();
+		let owner: T::AccountId = account("owner", 0, 0);
+
+		let (collection_id, schema, royalties) = setup_collection::<T>(creator.clone());
+		let _ = <Nft<T>>::create_collection(RawOrigin::Signed(creator.clone()).into(), collection_id.clone(), schema.clone(), Some(royalties.clone())).expect("created collection");
+		// all attributes max. length
+		let attributes = schema.iter().map(|_| NFTAttributeValue::String([1_u8; 140_usize].to_vec())).collect::<Vec<NFTAttributeValue>>();
```
Comment (aliXsed) [inline crml/nft/src/benchmarking.rs:102]: Could you already make schema like this in setup_collection?

Comment (aliXsed) [reply crml/nft/src/benchmarking.rs:102]: If 140 is a maximum len that should be respected in the module, could you please add 140 as a defined const in the module and "use" it here? If it should be coming from T::MaxAttributeLength::get(), then you can use T::MaxAttributeLength::get() directly here.

File: crml/nft/src/benchmarking.rs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+		let collection_id = setup_token::<T>(owner.clone());
+		let new_owner: T::AccountId = account("new_owner", 0, 0);
+		let token_id = T::TokenId::from(0_u32);
+
+	}: _(RawOrigin::Signed(owner.clone()), collection_id.clone(), token_id, new_owner.clone())
+	verify {
+		assert_eq!(<Nft<T>>::token_owner(&collection_id, token_id), new_owner);
+	}
+
+	burn {
```
Comment (aliXsed) [inline crml/nft/src/benchmarking.rs:122]: Looks like the expense of burn would be variable based on the number of tokens of a particular id that an owner has collected:
`			<CollectedTokens<T>>::mutate(&collection_id, current_owner, |tokens| {
				tokens.retain(|t| t != &token_id)
			});
`
Can you make the benchmark take that count into account or test a worst case/average scenario?

File: crml/nft/src/benchmarking.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+	cancel_sale {
+		let owner: T::AccountId = account("owner", 0, 0);
+		let collection_id = setup_token::<T>(owner.clone());
+		let token_id = T::TokenId::from(0_u32);
+		let payment_asset = 16_000;
+		let reserve_price = 1_000_000 * 10_000; // 1 million 4dp asset
+		let duration = T::BlockNumber::from(100_u32);
+
+		let _ = <Nft<T>>::auction(RawOrigin::Signed(owner.clone()).into(), collection_id.clone(), token_id, payment_asset, reserve_price, Some(duration)).expect("listed ok");
```
Comment (aliXsed) [inline crml/nft/src/benchmarking.rs:205]: Again here according to the following code:
`			<CollectedTokens<T>>::mutate(&collection_id, current_owner, |tokens| {
				tokens.retain(|t| t != &token_id)
			});
`
It seesm we should have a number of auctions scheduled to make it either the worst or average case scenario.

**Comment by aliXsed** [comment]
It lgtm! just left a few suggestions.

File: crml/nft/src/benchmarking.rs (hunk 3 above)
Comment (jordy25519) [reply crml/nft/src/benchmarking.rs:102]: needs a const here, added an assert as a workaround

**Pull request merged**